index.js

```diff
@@ -29,54 +29,104 @@
  */
 
 const path = require('path');
+const { Mutex } = require('async-mutex');
 
 function CaseSensitivePathsPlugin(options) {
   this.options = options || {};
   this.logger = this.options.logger || console;
   this.pathCache = new Map();
+
+  // stores a mutex for each path
+  //
+  // The goal:
+  //    We use pathCache to store already known filenames for each directory.
+  //
+  //    We want getFilenamesInDir() only call fs.readdir() once,
+  //    and then use cached result for further calls on same directory.
+  //
+  // The culprit:
+  //    1. There's a new directory `tippy`, which is not in pathCache.
+  //    2. getFilenamesInDir() gets called with `tippy`
+  //          check pathCache: cache is empty
+  //          call fs.readdir()
+  //    3. getFilenamesInDir() gets called again with `tippy`:
+  //          check pathCache: cache is empty (previous call to fs.readdir() is not finished)
+  //          call fs.readdir()
+  //    4. Before cache for `tippy` is written, more calls to getFilenamesInDir() with `tippy`,
+  //          causes as much as calls to fs.readdir()
+  // Thus we have to put a mutex for path `tippy`, when cache is not ready,
+  // to ensure we do not call fs.readdir() on same path multiple times.
+  this.visitedPathMutexRecord = new Map();
   this.reset();
 }
 
-CaseSensitivePathsPlugin.prototype.reset = function() {
+CaseSensitivePathsPlugin.prototype.reset = function () {
   this.pathCache = new Map();
+  this.visitedPathMutexRecord.clear();
   this.fsOperations = 0;
-  this.primed = false;
 };
 
-CaseSensitivePathsPlugin.prototype.getFilenamesInDir = function(dir, callback) {
+CaseSensitivePathsPlugin.prototype.getFilenamesInDir = function (dir, callback) {
   const that = this;
   const fs = this.compiler.inputFileSystem;
-  this.fsOperations += 1;
+
+  if (!this.visitedPathMutexRecord.has(dir)) {
+    this.visitedPathMutexRecord.set(dir, new Mutex());
+  }
 
   if (this.pathCache.has(dir)) {
+    if (that.options.debug) {
+      that.logger.log(
+        '[CaseSensitivePathsPlugin] Hit cache for directory',
+        dir,
+      );
+    }
     callback(this.pathCache.get(dir));
     return;
   }
-  if (this.options.debug) {
-    this.logger.log('[CaseSensitivePathsPlugin] Reading directory', dir);
-  }
 
-  fs.readdir(dir, (err, files) => {
-    if (err) {
+  this.visitedPathMutexRecord.get(dir).runExclusive(() => {
+    if (that.pathCache.has(dir)) {
```

Comment:
We may have several queued `runExclusive()` jobs when calling `getFilenamesInDir` with the same dir during a short while, after the first one writes the cache, the later ones should use the cached result.

Thus we have to check the cache first.